File: problems/206/solution1.cpp
// Accepted: 0ms (100.00%), 8.54MB (79.53%)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* cur = head;

        while (cur != nullptr) {
            ListNode* oldPrev = prev;

            prev = cur;
            cur = cur->next;

            prev->next = oldPrev;
        }

        return prev;
    }
};

File: problems/206/solution1.ts
// Accepted: 52ms (90.52%), 45.53MB (10.71%)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {
    let prev: ListNode | null = null;
    let cur: ListNode | null = head;

    while (cur !== null) {
        const oldPrev = prev;

        prev = cur;
        cur = cur.next;

        prev.next = oldPrev;
    }

    return prev;
};
